docs/development.rst

```diff
@@ -50,19 +50,277 @@ GitHub workflow
+    Date: Tue Mar 5 17:00:39 2019 +0000
+    1 file changed, 2 insertions(+), 2 deletions(-)
+    Successfully rebased and updated refs/heads/better-rebase-docs.
+
+Finally, after a successful rebase, you **must force-push**! If you try to
+push without specifying ``-f``, you will get a very confusing and misleading
+message::
+
+    $ git push origin better-rebase-docs
+    To github.com:mcveanlab/stdpopsim.git
+    ! [rejected]        better-rebase-docs -> better-rebase-docs (non-fast-forward)
+    error: failed to push some refs to 'git@github.com:mcveanlab/stdpopsim.git'
+    hint: Updates were rejected because the tip of your current branch is behind
+    hint: its remote counterpart. Integrate the remote changes (e.g.
+    hint: 'git pull ...') before pushing again.
+    hint: See the 'Note about fast-forwards' in 'git push --help' for details.
+
+**DO NOT LISTEN TO GIT IN THIS CASE!** Git is giving you is **terrible advice**
+which will mess up your branch. Instead, force-push::
+
+    $ git push -f origin better-rebase-docs
+    Counting objects: 4, done.
+    Delta compression using up to 4 threads.
+    Compressing objects: 100% (4/4), done.
+    Writing objects: 100% (4/4), 4.33 KiB | 1.44 MiB/s, done.
+    Total 4 (delta 2), reused 0 (delta 0)
+    remote: Resolving deltas: 100% (2/2), completed with 2 local objects.
+    To github.com:mcveanlab/stdpopsim.git
+     + 6b8a2a5...d033ffa better-rebase-docs -> better-rebase-docs (forced update)
+
+Success! We can check the history again to see if everything looks OK::
+
+    $  git log --decorate --oneline --graph
+
+    * d033ffa (HEAD -> better-rebase-docs, origin/better-rebase-docs) Add documentation for rebasing.
+    *   20fb835 (upstream/master) Merge pull request #22 from mcveanlab/port-tennyson
+    |\
+    | * b3d45ea (origin/port-tennyson, port-tennyson) Quickly port Tennesen et al model.
+    |/
+    *   79d26b4 Merge pull request #20 from andrewkern/fly_model
+    |
+
+This looks just right: we have one commit, pointing to the head of ``upstream/master``
+and have successfully squashed and rebased.
+
+------------------------
+When rebasing goes wrong
+------------------------
 
-       Rebasing is very powerful (and complicated!), so please see
-       `this guide <https://help.github.com/articles/about-git-rebase/>`_
-       for more information. Once you have rebased your topic branch, you
-       then need to "force push" ::
+Sometimes rebasing goes wrong, and you end up in a frustrating loop of making and
+undoing the same changes over and over again. In this case, it can be simplest to
+just make a diff of your current changes, and apply these in a single commit::
 
-        $ git push -f origin topic_branch_name
+    $ git diff upstream/master > changes.patch
+    $ git reset --hard upstream/master
+    $ patch -p1 < changes.patch
+    $ git commit -a
+    $ git push -f origin my_topic_branch
 
+This will completely remove **all** of your old commits, so beware!
```

Comment:
perhaps we should suggest doing this on a separate branch first, to check that it works ok? Like

```
git checkout -b test_branch upstream/master
patch -p1 < changes.patch
git commit -a
# check things work
git checkout my_topic_branch
git reset --hard test_branch
git push -f origin my_topic_branch 
```

Reply:
Yeah, good idea. This is very much the nuclear option, needs to be treated with care.